src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,804 @@
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	if (cd->process_enable[chan]) {
+		cd->process_enable[chan] = false;
+		IgoLibInit(cd->p_handle, &cd->igo_lib_config, &cd->config.igo_params);
+	}
+}
+
+static int32_t igo_nr_cmd_get_data(struct comp_dev *dev,
+				   struct sof_ipc_ctrl_data *cdata,
+				   int32_t max_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "igo_nr_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+		ret = comp_data_blob_get_cmd(cd->model_handler, cdata, max_size);
+		break;
+	default:
+		comp_err(dev, "igo_nr_cmd_get_data() error: invalid cdata->cmd", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int32_t igo_nr_cmd_get_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t j;
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_ENUM:
+		comp_info(dev, "igo_nr_cmd_get_value(), SOF_CTRL_CMD_ENUM index=%d", cdata->index);
+		switch (cdata->index) {
+		case IGO_NR_ONOFF_SWITCH:
+			for (j = 0; j < cdata->num_elems; j++)
+				cdata->chanv[j].value = cd->process_enable[j];
+			break;
+		default:
+			comp_err(dev, "igo_nr_cmd_get_value() error: invalid cdata->index %d",
+				 cdata->index);
+			ret = -EINVAL;
+			break;
+		}
+		break;
+	case SOF_CTRL_CMD_SWITCH:
+		for (j = 0; j < cdata->num_elems; j++) {
+			cdata->chanv[j].channel = j;
+			cdata->chanv[j].value = cd->process_enable[j];
+			comp_info(dev, "igo_nr_cmd_get_value(), channel = %u, value = %u",
+				  cdata->chanv[j].channel,
+				  cdata->chanv[j].value);
+		}
+		break;
+
+	default:
+		comp_err(dev, "igo_nr_cmd_get_value() error: invalid cdata->cmd %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int32_t igo_nr_cmd_set_data(struct comp_dev *dev,
+				   struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "igo_nr_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+		ret = comp_data_blob_set_cmd(cd->model_handler, cdata);
+		break;
+	default:
+		comp_err(dev, "igo_nr_cmd_set_data() error: invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int32_t igo_nr_set_chan(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	uint32_t val;
+	int32_t ch;
+	int32_t j;
+	int32_t ret = 0;
+
+	for (j = 0; j < cdata->num_elems; j++) {
+		ch = cdata->chanv[j].channel;
+		val = cdata->chanv[j].value;
+		comp_info(dev, "igo_nr_cmd_set_value(), channel = %d, value = %u", ch, val);
+		if (ch < 0 || ch >= SOF_IPC_MAX_CHANNELS) {
+			comp_err(dev, "igo_nr_cmd_set_value(), illegal channel = %d", ch);
+			return -EINVAL;
+		}
+
+		if (val)
+			igo_nr_set_chan_process(dev, ch);
+		else
+			igo_nr_set_chan_passthrough(dev, ch);
+	}
+
+	return ret;
```

Comment:
looks like we dont need ret at all. just remoe it and return 0?

Reply:
the ret was missing. I add it back. thanks!

Reply:
Yes you are right. Return 0 here is just fine. Thanks!